js/classes/zone.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * A Zone is a small portion of the Game World. 
+ * The Game Engine will load and deal with one Zone at a time.
+ * A Dimension (once an actual class, now simply an in-game concept) will contain several Zones.
+ * Zones should not be created willy-nilly - the only way you should actually construct a Zone is from the static getZones() method!
+ * 
+ * A Zone will define the minimum (upper left) and maximum (bottom right) point which is going to be visible to the Camera -
+ * anything outside this range will never be visible.
+ * A Zone will also contain a load method: this ensures that all assets are downloaded, all entities are placed (INCLUDING CHAD), etc.
+ * 
+ * @author Devin Peevy
+ */
+class Zone {
+
+    /**
+     * This constructor is PRIVATE should NOT be used outside the Zone class. To load a Zone, use the Zone.getZones() static method.
+     * This constructor can be used INSIDE THAT METHOD to dictate the min/max point and loadFunction.
+     * @param {Vector} minBlock The FIRST (top left) block which should be visible in this Zone. (NOT PIXEL COORDINATES)
+     * @param {Vector} maxBlock The LAST (bottom right) block which should be visible in this Zone. (NOT PIXEL COORDINATES)
+     * @param {function} loadFunction A function which is going to download all necessary assets and add all necessary entities into the zone.
+     */
+    constructor(minBlock, maxBlock, loadFunction) {
+        this.minBlock = minBlock;
+        this.maxBlock = maxBlock;
+        this.load = () => {
+            // Clear all entities from previous Zone.
+            GAME.clearEntities();
+            // Clear all images/music from previous Zone.
+            ASSET_MGR.clearCache();
+            loadFunction();
+        };
+    };
+
+    // DYNAMIC GETTERS:
+
+    /**
+     * @returns {Vector} The Min (upper left) block that is visible in this Zone.
+     */
+    get MIN_BLOCK() {
+        return this.minBlock;
+    };
+
+    /**
+     * @returns {Vector} The Max (bottom right) block that is visible in this Zone.
+     */
+    get MAX_BLOCK() {
+        return this.maxBlock;
+    };
+
+    /**
+     * @returns {Vector} The minimum (x, y) coordinate visible in this Zone.
+     */
+    get MIN_PT() {
+        return Vector.multiply(this.MIN_BLOCK, Block.SCALED_SIZE);
+    };
+
+    /**
+     * @returns {Vector} The maximum (x, y) coordinate visible in this Zone.
+     */
+    get MAX_PT() {
+        // I want to actually be able to SEE the final block (not just touch its top left corner). 
+        // Therefore:
+        const ones = new Vector(1, 1);
+        return Vector.multiply(Vector.add(this.MAX_BLOCK, ones), Block.SCALED_SIZE);
```

Comment:
I'm a little confused on what is happening here specifically. I understand that we want the max block to be shown on the screen fully however I am a bit confused on how the vector of ones factors into this. Wouldn't that be a minimum point? @B1GD3VD0G What does the vector of MAX_BLOCK represent and MIN_BLOCK? If we aren't referring to coordinates on the block what does the vector represent? Vector is a combination of x and y right? So what are the x and y values used for a block that is being referred to?

Reply:
MAX_BLOCK and MIN_BLOCK represent the first (top/left) and last(bottom/right) block that is ENTIRELY WITHIN THE BOUNDS of the zone - a zone is by design, rectangular. It is easy to calculate the game coordinates of the top left corner, using Trae's Vector.blockToGameCoordinates (or similar title) method, or as I did here, multiplying by Block.SCALED_SIZE. However, the top left corner of my MAX_BLOCK is not enough - i want to actually see the whole thing. So instead, what I calculate is the top left coordinate of the block one below, and one to the right of it.

Reply:
So we are just looking at the final point on the screen which ends up in the bottom-right corner of one block how you calculate it and above the true-final block. Because if you didn't calculate it like that then it would be in the top-left of the actual final block. Is that correct?

Reply:
I think so? Take this example: the last box is (100,100). This converted into game coordinates is (6400,6400) (a block is 64 pixels rn, I believe).

If we are to set max point to (6400,6400), we would never see any of the boxes in row OR column 100, as this is the coordinate of the top left corner of the block(100,100)

Therefore, instead of converting MAX_BLOCK to game coordinates immediately, I add one to x and y, and set MAX_PT to be the top left coordinate of block(101,101). Then I actually see row and column 100 in my zone.